File: scripts/task_loop_lease.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
[Lease Manager] task-loop 独立租约锁模块 (Python 标准库零依赖)

统一数据结构:
{
  "schema_version": 5,
  "lease_id": "<uuid>",
  "resource": "sessions" | "todo" | "policy",
  "owner": {
    "vendor": "antigravity",
    "session_id": "..."
  },
  "holder_thread_id": "...",
  "acquired_at": "<iso>",
  "heartbeat_at": "<iso>",
  "expires_at": "<iso>"
}

核心方法:
- acquire(resource, owner, options)
- renew(lease_id, owner, options)
- release(lease_id, owner, options)
- expire(resource, options)
- inspect(resource, options)

保证并发安全:
- 基于底层文件系统独占原子排他互斥锁 (.lock w/ O_CREAT | O_EXCL);
- 0 corrupt JSON, 0 duplicate owner, 0 lost update.
"""

import json
import os
import random
import sys
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path

# 保证当前 scripts 目录在 sys.path 中
_CURRENT_DIR = Path(__file__).resolve().parent
if str(_CURRENT_DIR) not in sys.path:
    sys.path.insert(0, str(_CURRENT_DIR))

from task_loop_state import atomic_write_json, read_json, normalize_vendor

DEFAULT_LEASE_FILE = _CURRENT_DIR.parent / ".agents" / "task-loop" / "lease.json"
DEFAULT_TTL_MS = 5 * 60 * 1000  # 5 分钟
DEFAULT_LOCK_TIMEOUT_S = 10.0
STALE_LOCK_THRESHOLD_S = 10.0


def _now_iso():
    return datetime.now(timezone.utc).isoformat()


def _parse_iso_ms(iso_str):
    try:
        dt = datetime.fromisoformat(iso_str.replace("Z", "+00:00"))
        return dt.timestamp() * 1000.0
    except Exception:
        return 0.0


def _iso_from_timestamp_ms(ts_ms):
    return datetime.fromtimestamp(ts_ms / 1000.0, timezone.utc).isoformat()


class LeaseSchemaError(ValueError):
    def __init__(self, message="Lease Schema 校验未通过"):
        super().__init__(message)
        self.code = "TL_STATE_INVALID_SCHEMA"


class LockTimeoutError(TimeoutError):
    def __init__(self, message="获取排他文件锁超时"):
        super().__init__(message)
        self.code = "TL_LOCK_TIMEOUT"


def validate_lease(lease, allow_released=False):
    """校验 Lease 对象结构 (基于 Schema 契约)"""
    if not isinstance(lease, dict):
        raise LeaseSchemaError("Lease 根节点必须为对象")
    if allow_released and lease.get("lease_id") is None:
        return True
    if not lease.get("lease_id") or not isinstance(lease["lease_id"], str):
        raise LeaseSchemaError("Lease 缺少有效 lease_id")
    if not lease.get("resource") or not isinstance(lease["resource"], str):
        raise LeaseSchemaError("Lease 缺少有效 resource 字段")
    owner = lease.get("owner")
    if not isinstance(owner, dict) or not owner.get("vendor") or not owner.get("session_id"):
        raise LeaseSchemaError("Lease 缺少有效 owner 对象 (vendor 与 session_id)")
    if not lease.get("acquired_at") or not lease.get("expires_at"):
        raise LeaseSchemaError("Lease 缺少有效 acquired_at 或 expires_at 字段")
    return True


def acquire_file_lock(lock_file, timeout_s=DEFAULT_LOCK_TIMEOUT_S):
    """获取底层文件原子排他锁 (.lock 带有 O_CREAT | O_EXCL 标志)"""
    lock_path = Path(lock_file).resolve()
    lock_path.parent.mkdir(parents=True, exist_ok=True)
    start = time.time()

    while time.time() - start < timeout_s:
        try:
            fd = os.open(str(lock_path), os.O_CREAT | os.O_EXCL | os.O_RDWR)
            def unlock():
                try:
                    os.close(fd)
                except Exception:
                    pass
                for _ in range(5):
                    try:
                        if lock_path.exists():
                            lock_path.unlink()
                        break
                    except Exception:
                        time.sleep(0.005)
            return unlock
        except (FileExistsError, PermissionError, OSError):
            try:
                if lock_path.exists():
                    mtime = lock_path.stat().st_mtime
                    if time.time() - mtime > STALE_LOCK_THRESHOLD_S:
                        try:
                            lock_path.unlink()
                        except Exception:
                            pass
            except Exception:
                pass
            time.sleep(0.01 + random.uniform(0.005, 0.02))
        except Exception as e:
            raise e

    raise LockTimeoutError(f"获取文件锁超时 ({timeout_s}s): {lock_file}")


def with_file_lock(lease_file, fn, timeout_s=DEFAULT_LOCK_TIMEOUT_S):
    """在底层独占锁保护下执行临界区操作"""
    lock_file = f"{lease_file}.lock"
    unlock = acquire_file_lock(lock_file, timeout_s)
    try:
        return fn()
    finally:
        unlock()


class LeaseManager:
    def __init__(self, lease_path=None, ttl_ms=DEFAULT_TTL_MS, lock_timeout_s=DEFAULT_LOCK_TIMEOUT_S):
        self.lease_path = Path(lease_path).resolve() if lease_path else DEFAULT_LEASE_FILE
        self.default_ttl_ms = ttl_ms
        self.lock_timeout_s = lock_timeout_s

    def _resolve_file(self, options=None):
        options = options or {}
        if options.get("lease_path") or options.get("leasePath"):
            return Path(options.get("lease_path") or options.get("leasePath")).resolve()
        return self.lease_path

    def _resolve_ttl_ms(self, options=None):
        options = options or {}
        if options.get("ttl_ms") is not None:
            return options["ttl_ms"]
        if options.get("ttlMs") is not None:
            return options["ttlMs"]
        if options.get("lease_minutes") is not None:
            return options["lease_minutes"] * 60 * 1000
        if options.get("leaseMinutes") is not None:
            return options["leaseMinutes"] * 60 * 1000
        return self.default_ttl_ms

    def acquire(self, resource, owner, options=None):
        """抢占租约锁 (Acquire)"""
        options = options or {}
        if not resource:
            raise ValueError("acquire: resource 不能为空")
        if not isinstance(owner, dict) or not owner.get("vendor") or not owner.get("session_id"):
            raise ValueError("acquire: owner 必须包含 vendor 和 session_id")

        norm_vendor = normalize_vendor(owner.get("vendor")) or owner.get("vendor")
        norm_owner = {"vendor": norm_vendor, "session_id": str(owner.get("session_id"))}
        target_file = self._resolve_file(options)
        ttl_ms = self._resolve_ttl_ms(options)
        timeout_s = options.get("lock_timeout_s") or options.get("lockTimeoutMs", 0) / 1000.0 or self.lock_timeout_s

        def _critical():
            existing = read_json(target_file)
            now_ms = time.time() * 1000.0
            now_str = _now_iso()

            if isinstance(existing, dict) and existing.get("lease_id") and existing.get("expires_at"):
                exp_ms = _parse_iso_ms(existing["expires_at"])
                is_expired = now_ms >= exp_ms

                if not is_expired and (existing.get("resource") == resource or not existing.get("resource")):
                    ex_owner = existing.get("owner")
                    is_same_owner = (
                        isinstance(ex_owner, dict)
                        and ex_owner.get("vendor") == norm_owner["vendor"]
                        and str(ex_owner.get("session_id")) == norm_owner["session_id"]
                    )

                    if is_same_owner:
                        # 重入续期
                        existing["heartbeat_at"] = now_str
                        existing["expires_at"] = _iso_from_timestamp_ms(now_ms + ttl_ms)
                        validate_lease(existing)
                        atomic_write_json(target_file, existing)
                        return {
                            "success": True,
                            "acquired": True,
                            "reentered": True,
                            "lease_id": existing["lease_id"],
                            "lease": existing,
                        }

                    return {
                        "success": False,
                        "acquired": False,
                        "reason": "LEASE_HELD_BY_OTHER",
                        "active_lease": existing,
                    }

            # 授予新租约
            lease_id = str(uuid.uuid4())
            new_lease = {
                "schema_version": 5,
                "lease_id": lease_id,
                "resource": resource,
                "owner": norm_owner,
                "holder_thread_id": norm_owner["session_id"],
                "acquired_at": now_str,
                "heartbeat_at": now_str,
                "expires_at": _iso_from_timestamp_ms(now_ms + ttl_ms),
            }

            validate_lease(new_lease)
            atomic_write_json(target_file, new_lease)

            return {
                "success": True,
                "acquired": True,
                "reentered": False,
                "lease_id": lease_id,
                "lease": new_lease,
            }

        return with_file_lock(target_file, _critical, timeout_s)

    def renew(self, lease_id, owner=None, options=None):
        """续期/心跳 (Renew)"""
        options = options or {}
        if not lease_id:
            raise ValueError("renew: lease_id 不能为空")
        target_file = self._resolve_file(options)
        ttl_ms = self._resolve_ttl_ms(options)
        timeout_s = options.get("lock_timeout_s") or options.get("lockTimeoutMs", 0) / 1000.0 or self.lock_timeout_s

        def _critical():
            existing = read_json(target_file)
            if not isinstance(existing, dict) or existing.get("lease_id") != lease_id:
                return {"success": False, "renewed": False, "reason": "LEASE_NOT_HELD"}

            if owner:
                norm_vendor = normalize_vendor(owner.get("vendor")) or owner.get("vendor")
                ex_owner = existing.get("owner", {})
                if ex_owner.get("vendor") != norm_vendor or str(ex_owner.get("session_id")) != str(owner.get("session_id")):
                    return {"success": False, "renewed": False, "reason": "OWNER_MISMATCH"}

            now_ms = time.time() * 1000.0
            existing["heartbeat_at"] = _now_iso()
            existing["expires_at"] = _iso_from_timestamp_ms(now_ms + ttl_ms)

            validate_lease(existing)
            atomic_write_json(target_file, existing)

            return {"success": True, "renewed": True, "lease": existing}

        return with_file_lock(target_file, _critical, timeout_s)

    def release(self, lease_id, owner=None, options=None):
        """释放租约 (Release)"""
        options = options or {}
        force = options.get("force", False)
        if not lease_id and not force:
            raise ValueError("release: lease_id 不能为空 (除非指定 force=True)")

        target_file = self._resolve_file(options)
        timeout_s = options.get("lock_timeout_s") or options.get("lockTimeoutMs", 0) / 1000.0 or self.lock_timeout_s

        def _critical():
            existing = read_json(target_file)
            if not isinstance(existing, dict) or (not force and existing.get("lease_id") != lease_id):
                return {"success": False, "released": False, "reason": "LEASE_NOT_HELD"}

            if not force and owner:
                norm_vendor = normalize_vendor(owner.get("vendor")) or owner.get("vendor")
                ex_owner = existing.get("owner", {})
                if ex_owner.get("vendor") != norm_vendor or str(ex_owner.get("session_id")) != str(owner.get("session_id")):
                    return {"success": False, "released": False, "reason": "OWNER_MISMATCH"}

            released_lease = {
                "schema_version": 5,
                "lease_id": None,
                "resource": existing.get("resource", "sessions"),
                "owner": None,
                "holder_thread_id": None,
                "acquired_at": None,
                "heartbeat_at": None,
                "expires_at": None,
            }

            atomic_write_json(target_file, released_lease)
            return {"success": True, "released": True}

        return with_file_lock(target_file, _critical, timeout_s)

    def expire(self, resource=None, options=None):
        """强制使租约过期 (Expire)"""
        options = options or {}
        target_file = self._resolve_file(options)
        timeout_s = options.get("lock_timeout_s") or options.get("lockTimeoutMs", 0) / 1000.0 or self.lock_timeout_s

        def _critical():
            existing = read_json(target_file)
            if not isinstance(existing, dict) or not existing.get("lease_id"):
                return {"success": True, "expired": False, "reason": "NO_ACTIVE_LEASE"}
            if resource and existing.get("resource") and existing.get("resource") != resource:
                return {"success": False, "expired": False, "reason": "RESOURCE_MISMATCH"}

            existing["expires_at"] = "1970-01-01T00:00:00+00:00"
            atomic_write_json(target_file, existing)
            return {"success": True, "expired": True, "lease": existing}

        return with_file_lock(target_file, _critical, timeout_s)

    def inspect(self, resource=None, options=None):
        """查看当前租约状态 (Inspect)"""
        options = options or {}
        target_file = self._resolve_file(options)
        existing = read_json(target_file)

        if not isinstance(existing, dict) or not existing.get("lease_id") or not existing.get("expires_at"):
            return {"active": False, "is_expired": False, "lease": None}
        if resource and existing.get("resource") and existing.get("resource") != resource:
            return {"active": False, "is_expired": False, "lease": None, "other_resource": existing.get("resource")}

        now_ms = time.time() * 1000.0
        exp_ms = _parse_iso_ms(existing["expires_at"])
        is_expired = now_ms >= exp_ms

        return {
            "active": not is_expired,
            "is_expired": is_expired,
            "lease": existing,
        }


_default_manager = LeaseManager()


def acquire(resource, owner, options=None):
    return _default_manager.acquire(resource, owner, options)


def renew(lease_id, owner=None, options=None):
    return _default_manager.renew(lease_id, owner, options)


def release(lease_id, owner=None, options=None):
    return _default_manager.release(lease_id, owner, options)


def expire(resource=None, options=None):
    return _default_manager.expire(resource, options)


def inspect(resource=None, options=None):
    return _default_manager.inspect(resource, options)

File: scripts/task_loop_lease.js
#!/usr/bin/env node
/**
 * [Lease Manager] task-loop 独立租约锁模块 (Node.js)
 *
 * 统一数据结构:
 * {
 *   "schema_version": 5,
 *   "lease_id": "<uuid>",
 *   "resource": "sessions" | "todo" | "policy",
 *   "owner": {
 *     "vendor": "antigravity",
 *     "session_id": "..."
 *   },
 *   "holder_thread_id": "...",
 *   "acquired_at": "<iso>",
 *   "heartbeat_at": "<iso>",
 *   "expires_at": "<iso>"
 * }
 *
 * 核心方法:
 * - acquire(resource, owner, options)
 * - renew(leaseId, owner, options)
 * - release(leaseId, owner, options)
 * - expire(resource, options)
 * - inspect(resource, options)
 *
 * 保证并发安全:
 * - 基于底层文件系统独占原子排他互斥锁 (.lock w/ O_EXCL 'wx' 标志);
 * - 0 corrupt JSON, 0 duplicate owner, 0 lost update.
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { atomicWriteJson, readJson, normalizeVendor } = require('./task_loop_state');

const DEFAULT_LEASE_FILE = path.resolve(__dirname, '..', '.agents', 'task-loop', 'lease.json');
const DEFAULT_TTL_MS = 5 * 60 * 1000; // 5 分钟
const DEFAULT_LOCK_TIMEOUT_MS = 10000; // 10 秒
const STALE_LOCK_THRESHOLD_MS = 10000; // 10 秒视为失效僵死锁

function generateUuid() {
  if (crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

function nowIso() {
  return new Date().toISOString();
}

function sleepSync(ms) {
  try {
    const buf = new Int32Array(new SharedArrayBuffer(4));
    Atomics.wait(buf, 0, 0, ms);
  } catch (_) {
    const end = Date.now() + ms;
    while (Date.now() < end) {}
  }
}

/**
 * 校验 Lease 对象结构 (基于 Schema 契约)
 */
function validateLease(lease, { allowReleased = false } = {}) {
  if (!lease || typeof lease !== 'object' || Array.isArray(lease)) {
    const err = new Error('Lease 根节点必须为对象');
    err.code = 'TL_STATE_INVALID_SCHEMA';
    throw err;
  }
  if (allowReleased && lease.lease_id === null) {
    return true;
  }
  if (!lease.lease_id || typeof lease.lease_id !== 'string') {
    const err = new Error('Lease 缺少有效 lease_id');
    err.code = 'TL_STATE_INVALID_SCHEMA';
    throw err;
  }
  if (!lease.resource || typeof lease.resource !== 'string') {
    const err = new Error('Lease 缺少有效 resource 字段');
    err.code = 'TL_STATE_INVALID_SCHEMA';
    throw err;
  }
  if (!lease.owner || typeof lease.owner !== 'object' || !lease.owner.vendor || !lease.owner.session_id) {
    const err = new Error('Lease 缺少有效 owner 对象 (vendor 与 session_id)');
    err.code = 'TL_STATE_INVALID_SCHEMA';
    throw err;
  }
  if (!lease.acquired_at || !lease.expires_at) {
    const err = new Error('Lease 缺少有效 acquired_at 或 expires_at 字段');
    err.code = 'TL_STATE_INVALID_SCHEMA';
    throw err;
  }
  return true;
}

/**
 * 获取底层文件原子排他锁 (.lock 带有 'wx' 标志)
 */
function acquireFileLock(lockFile, timeoutMs = DEFAULT_LOCK_TIMEOUT_MS) {
  const start = Date.now();
  const dir = path.dirname(lockFile);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  while (Date.now() - start < timeoutMs) {
    try {
      const fd = fs.openSync(lockFile, 'wx');
      return () => {
        try {
          fs.closeSync(fd);
        } catch (_) {}
        for (let i = 0; i < 5; i++) {
          try {
            if (fs.existsSync(lockFile)) fs.unlinkSync(lockFile);
            break;
          } catch (_) {
            sleepSync(5);
          }
        }
      };
    } catch (err) {
      if (err.code === 'EEXIST' || err.code === 'EPERM' || err.code === 'EACCES') {
        try {
          if (fs.existsSync(lockFile)) {
            const stat = fs.statSync(lockFile);
            if (Date.now() - stat.mtimeMs > STALE_LOCK_THRESHOLD_MS) {
              try { fs.unlinkSync(lockFile); } catch (_) {}
            }
          }
        } catch (_) {}
        sleepSync(10 + Math.floor(Math.random() * 20));
        continue;
      }
      throw err;
    }
  }
  const timeoutErr = new Error(`获取文件锁超时 (${timeoutMs}ms): ${lockFile}`);
  timeoutErr.code = 'TL_LOCK_TIMEOUT';
  throw timeoutErr;
}

/**
 * 在底层独占锁保护下执行临界区操作
 */
function withFileLock(leaseFile, fn, timeoutMs = DEFAULT_LOCK_TIMEOUT_MS) {
  const lockFile = `${leaseFile}.lock`;
  const unlock = acquireFileLock(lockFile, timeoutMs);
  try {
    return fn();
  } finally {
    unlock();
  }
}

class LeaseManager {
  constructor(options = {}) {
    this.leasePath = options.leasePath ? path.resolve(options.leasePath) : DEFAULT_LEASE_FILE;
    this.defaultTtlMs = options.ttlMs || DEFAULT_TTL_MS;
    this.lockTimeoutMs = options.lockTimeoutMs || DEFAULT_LOCK_TIMEOUT_MS;
  }

  _resolveFile(options = {}) {
    return options.leasePath ? path.resolve(options.leasePath) : this.leasePath;
  }

  _resolveTtl(options = {}) {
    if (options.ttlMs) return options.ttlMs;
    if (options.leaseMinutes) return options.leaseMinutes * 60 * 1000;
    return this.defaultTtlMs;
  }

  /**
   * 抢占租约锁 (Acquire)
   */
  acquire(resource, owner, options = {}) {
    if (!resource) throw new Error('acquire: resource 不能为空');
    if (!owner || !owner.vendor || !owner.session_id) {
      throw new Error('acquire: owner 必须包含 vendor 和 session_id');
    }

    const normVendor = normalizeVendor(owner.vendor) || owner.vendor;
    const normOwner = { vendor: normVendor, session_id: String(owner.session_id) };
    const targetFile = this._resolveFile(options);
    const ttlMs = this._resolveTtl(options);

    return withFileLock(targetFile, () => {
      const existing = readJson(targetFile);
      const now = Date.now();
      const nowStr = new Date(now).toISOString();

      if (existing && existing.lease_id && existing.expires_at) {
        const expiresTime = new Date(existing.expires_at).getTime();
        const isExpired = now >= expiresTime;

        // 若租约尚未过期且资源匹配
        if (!isExpired && (existing.resource === resource || !existing.resource)) {
          const isSameOwner = existing.owner &&
            existing.owner.vendor === normOwner.vendor &&
            existing.owner.session_id === normOwner.session_id;

          if (isSameOwner) {
            // 同一 owner 可重入并续期
            const newExpires = new Date(now + ttlMs).toISOString();
            existing.heartbeat_at = nowStr;
            existing.expires_at = newExpires;
            validateLease(existing);
            atomicWriteJson(targetFile, existing);
            return {
              success: true,
              acquired: true,
              reentered: true,
              lease_id: existing.lease_id,
              lease: existing
            };
          }

          // 其他会话持有有效租约，无法抢占
          return {
            success: false,
            acquired: false,
            reason: 'LEASE_HELD_BY_OTHER',
            active_lease: existing
          };
        }
      }

      // 未被占用或已过期: 授予新租约
      const leaseId = generateUuid();
      const newLease = {
        schema_version: 5,
        lease_id: leaseId,
        resource: resource,
        owner: normOwner,
        holder_thread_id: normOwner.session_id,
        acquired_at: nowStr,
        heartbeat_at: nowStr,
        expires_at: new Date(now + ttlMs).toISOString()
      };

      validateLease(newLease);
      atomicWriteJson(targetFile, newLease);

      return {
        success: true,
        acquired: true,
        reentered: false,
        lease_id: leaseId,
        lease: newLease
      };
    }, options.lockTimeoutMs || this.lockTimeoutMs);
  }

  /**
   * 续期/心跳 (Renew)
   */
  renew(leaseId, owner, options = {}) {
    if (!leaseId) throw new Error('renew: leaseId 不能为空');
    const targetFile = this._resolveFile(options);
    const ttlMs = this._resolveTtl(options);

    return withFileLock(targetFile, () => {
      const existing = readJson(targetFile);
      if (!existing || existing.lease_id !== leaseId) {
        return { success: false, renewed: false, reason: 'LEASE_NOT_HELD' };
      }

      if (owner) {
        const normVendor = normalizeVendor(owner.vendor) || owner.vendor;
        if (existing.owner.vendor !== normVendor || existing.owner.session_id !== String(owner.session_id)) {
          return { success: false, renewed: false, reason: 'OWNER_MISMATCH' };
        }
      }

      const now = Date.now();
      existing.heartbeat_at = new Date(now).toISOString();
      existing.expires_at = new Date(now + ttlMs).toISOString();

      validateLease(existing);
      atomicWriteJson(targetFile, existing);

      return { success: true, renewed: true, lease: existing };
    }, options.lockTimeoutMs || this.lockTimeoutMs);
  }

  /**
   * 释放租约 (Release)
   */
  release(leaseId, owner, options = {}) {
    if (!leaseId && !options.force) throw new Error('release: leaseId 不能为空 (除非指定 force: true)');
    const targetFile = this._resolveFile(options);

    return withFileLock(targetFile, () => {
      const existing = readJson(targetFile);
      if (!existing || (!options.force && existing.lease_id !== leaseId)) {
        return { success: false, released: false, reason: 'LEASE_NOT_HELD' };
      }

      if (!options.force && owner) {
        const normVendor = normalizeVendor(owner.vendor) || owner.vendor;
        if (existing.owner.vendor !== normVendor || existing.owner.session_id !== String(owner.session_id)) {
          return { success: false, released: false, reason: 'OWNER_MISMATCH' };
        }
      }

      const releasedLease = {
        schema_version: 5,
        lease_id: null,
        resource: existing.resource || 'sessions',
        owner: null,
        holder_thread_id: null,
        acquired_at: null,
        heartbeat_at: null,
        expires_at: null
      };

      atomicWriteJson(targetFile, releasedLease);
      return { success: true, released: true };
    }, options.lockTimeoutMs || this.lockTimeoutMs);
  }

  /**
   * 强制使租约过期 (Expire)
   */
  expire(resource, options = {}) {
    const targetFile = this._resolveFile(options);
    return withFileLock(targetFile, () => {
      const existing = readJson(targetFile);
      if (!existing || !existing.lease_id) {
        return { success: true, expired: false, reason: 'NO_ACTIVE_LEASE' };
      }
      if (resource && existing.resource && existing.resource !== resource) {
        return { success: false, expired: false, reason: 'RESOURCE_MISMATCH' };
      }

      existing.expires_at = '1970-01-01T00:00:00.000Z';
      atomicWriteJson(targetFile, existing);
      return { success: true, expired: true, lease: existing };
    }, options.lockTimeoutMs || this.lockTimeoutMs);
  }

  /**
   * 查看当前租约状态 (Inspect)
   */
  inspect(resource, options = {}) {
    const targetFile = this._resolveFile(options);
    const existing = readJson(targetFile);
    if (!existing || !existing.lease_id || !existing.expires_at) {
      return { active: false, is_expired: false, lease: null };
    }
    if (resource && existing.resource && existing.resource !== resource) {
      return { active: false, is_expired: false, lease: null, other_resource: existing.resource };
    }

    const now = Date.now();
    const expiresTime = new Date(existing.expires_at).getTime();
    const isExpired = now >= expiresTime;

    return {
      active: !isExpired,
      is_expired: isExpired,
      lease: existing
    };
  }
}

// 模块级单例工厂与快捷方法
const defaultManager = new LeaseManager();

module.exports = {
  LeaseManager,
  validateLease,
  acquireFileLock,
  withFileLock,
  acquire: (resource, owner, options) => defaultManager.acquire(resource, owner, options),
  renew: (leaseId, owner, options) => defaultManager.renew(leaseId, owner, options),
  release: (leaseId, owner, options) => defaultManager.release(leaseId, owner, options),
  expire: (resource, options) => defaultManager.expire(resource, options),
  inspect: (resource, options) => defaultManager.inspect(resource, options)
};
